packages/grafana-data/src/types/datasource.ts

```diff
@@ -25,6 +25,15 @@ export type DataSourceQueryType<DSType> = DSType extends DataSourceApi<infer TQu
 // Utility type to extract the options type TOptions from a class extending DataSourceApi<TQuery, TOptions>
 export type DataSourceOptionsType<DSType> = DSType extends DataSourceApi<any, infer TOptions> ? TOptions : never;
 
+export type QueryRelatedDataProviders = {
+  [name in string]: QueryRelatedDataProvider<any>;
+};
+
+export interface QueryRelatedDataProvider<TData> {
+  setRequest(request: DataQueryRequest): void;
```

Comment:
Why is the setRequest part of the API? Does it make sense to change it later on after the provider is returned? Otherwise I would assume it gets binded with the provider when created as client already specified the request when calling `getQueryRelatedDataProviders`. Changing it later seems like it could create undefined results as you don't know what other things were set up in the `getQueryRelatedDataProviders` method.

Reply:
Request object is actually binded to the provider, but I think it doesn't need to be a part of the API and could be provided in the constructor. The current logic is that if the related data is not available for the given query/request then no provider is returned. When it's available it's stored and kept until getData is called. 

Instead we could always return a provider regardless to the query (so it'd static for each data source) and pass `request` object to `getData(request)`. `getData` can return `undefined` if nothing is available for the given query. That would actually make the provider stateless so it could be a simple single function.

Reply:
I think the current version makes sense just that this method seems unnecessary/dangerous. Like this is how it works noe if I am reading this right:
```
const providers = datasource.getQueryRelatedDataProviders(request)
if (Object.values(providers).length) {
   providers['something'].getData()
}
```
So it makes sense, if there are nothing related to that query it just does not return any providers. If there is those providers already should be bound to the request. Doing something like this:
```
const providers = datasource.getQueryRelatedDataProviders(request)
if (Object.values(providers).length) {
   providers['something'].setRequest(someDifferentRequest)
   providers['something'].getData() // <--- what will this do after we set different request?
}
```
does not seem to make sense to me.

Having stateless provider also does not make sense to me because provider represents a query that can be run, so it is a piece of state. And we want it to be stateful probably because we want to inspect it and run it on demand (and as mentioned in other comment probably be able to get some metadata out of it). If it was `getData(request)` then this method would need to return some object with properties and run method which would just add one layer that isn't necessary.

Reply:
`setRequest(someDifferentRequest)` method is called inside `getQueryRelatedDataProviders()` when the provider is created (so it's not passed or called twice like in the snippet above). I somehow wanted to indicate that this type of provider needs to know the request and it has to be provided/implemented somehow. 

But since it's called internally and we don't have any generic factory for providers or anything like that it's unnecessary. `getQueryRelatedDataProviders` will use the constructor of a specific class to provide the request (so it can be in the constructor).

[x] remove setRequest method (2ae3b273bff22e4042d9b7c22d62c95c754e1ad0)